gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+    Args:
+         dummy_loader: Torch dataloader representing
+             the knots of the linear surrogate function.
+             The response variable can be anything.
+         model: Torch model
+    Returns
+    -------
+        results : Class 1 probabilities of the model evaluated on
+            the dummy_loader
+    """
+    model.eval()
+    featdim = next(iter(dummy_loader))[0].shape[1]
+    # featdim = dummy_loader.dataset.tensors[0].shape[1]
+    preds = []
+    for i, data in enumerate(dummy_loader, 0):
+        inputs, labels = data
+        inputs = torch.reshape(inputs, (-1, featdim)).float()
+        outputs = model(inputs)
+        preds.append(F.softmax(outputs, 1)[:, 1])
+    tmp = [p.detach().numpy() for p in preds]
+    results = np.concatenate(tmp, axis=0)
+    return results
+
+
+def get_persistent_pairs(
+    results: np.ndarray, connections: np.ndarray, expansion: int = 6
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    A helper file for topological regularizer
+    Given Class 1 probabilities and node connectivity information,
+    returns the persistent pairs of the function
+    associated to filtering according to the numerical values
+    Args:
+        results: an n-vector of class 1 probabilities
+        connections: an n by 2 numpy array describing pairwise connections
+        among the entries of results
+        expansion: The size of the clique expansion when computing homology
+    Returns
+    -------
+        out_pers: filtration values of the critical pairs
+        persts: the indices of the nodes corresponding to out_pers
+    """
+    st = SimplexTree()
+    for t in range(len(results)):
+        st.insert([t], filtration=results[t])
+    for i in range(connections.shape[0]):
+        in1 = connections[i, 0]
+        in2 = connections[i, 1]
+        fil = max(results[in1], results[in2])
+        st.insert([in1, in2], filtration=fil)
+    st.expansion(expansion)
+    st.compute_persistence()
+    tmp = st.persistence_pairs()
+    persistences = np.zeros([len(tmp), 2])
+    pers_indices = np.zeros([len(tmp), 2])
+    odd_one_out = 0
+    for i in range(len(tmp)):
+        inds1 = tmp[i][0]
+        inds2 = tmp[i][1]
+        if len(inds2) == 0:
+            odd_one_out = i
+            continue
+        for index in range(len(inds1)):
+            value = results[inds1[index]]
+            if value > persistences[i, 0]:
+                persistences[i, 0] = value
+                pers_indices[i, 0] = inds1[index]
+        for index in range(len(inds2)):
+            value = results[inds2[index]]
+            if value > persistences[i, 1]:
+                persistences[i, 1] = value
+                pers_indices[i, 1] = inds2[index]
+    out_pers = np.delete(persistences, odd_one_out, axis=0)
+    out_pers_ind = np.delete(pers_indices, odd_one_out, axis=0)
+    persts = out_pers_ind.astype(int)
+    return (out_pers, persts)
+
+
+def compute_critical_points(
```

Comment:
if this helper function is not meant to be used directly by the user, then please start its name with `_`

Reply:
done